rosbag2_storage/include/rosbag2_storage/storage_interfaces/base_read_interface.hpp

```diff
@@ -30,7 +30,7 @@ class ROSBAG2_STORAGE_PUBLIC BaseReadInterface
 public:
   virtual ~BaseReadInterface() = default;
 
-  virtual bool has_next() const = 0;
+  virtual bool has_next() = 0;
```

Comment:
We need to actually change the internal state of the storage to correctly ask whether there is a next entry. In the followup PR, we'll do this the following way (here, it's done as simple as possible):

- We have a `read_statement_` member which needs to be initialized somewhere. This is either done in `read_next()` or in `has_next()` if it is not already initialized. This read statement then gets reused for every read to allow fast reading. If we don't do this, `has_next()` would fail if `read_next()` has never been called, which is awkward.
- In order to allow the function to be `const` this would imply we'd have to make the read_statement (and by transitivity a lot of the other state of the storage) mutable, which is also not very nice. So we thought changing to non-`const` here is simpler.

If you think the `const` is worth it to signify that it shouldn't matter how often this function is called, all other function calls will result in the same results (i.e. the state as it is visible to users of this class is unchanged), we can change it back and make the necessary changes to mutable state.
If it's okay, I'd want to do this in a follow-up PR to avoid rebase hell for the other PRs because there are a lot of changes to the implementation in the other open PRs.

Reply:
Regarding the `const`ness: I agree that having `const` might improve the interface from a users point of view (for the sqlite implementation it probably would not be too hard to do so). However, this imposes a big limitation on the internal implementation of **all** future storage plugins (basically a lot of member need to be `mutable`).
One could also argue that the information whether more data is available might be expensive to obtain and thus should be considered an important part of the internal state.

Also if `has_next()` is `const` what about `read_next`? The arguments for/ against constness are pretty much the same.

To distinguish between read and write access we have the `ReadOnlyInterface` and `ReadWriteInterface`.